File: assignment2/a.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <time.h>

void vector_scalar_addition(float *vector, float scalar, int size, int num_threads) {
    // Set the number of threads
    omp_set_num_threads(num_threads);

    // Parallel vector scalar addition
    #pragma omp parallel for private(scalar)
    for (int i = 0; i < size; i++) {
        vector[i] += scalar;
    }
}

int main() {
    int sizes[] = {1000, 10000, 100000, 1000000, 10000000}; // Different vector sizes to test
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);
    int num_threads[] = {1, 2, 4, 8}; // Different thread counts to test
    int num_threads_count = sizeof(num_threads) / sizeof(num_threads[0]);

    float scalar = 5.0;

    for (int t = 0; t < num_threads_count; t++) {
        printf("\nPerformance with %d threads:\n", num_threads[t]);

        for (int s = 0; s < num_sizes; s++) {
            int size = sizes[s];
            float *vector = (float *)malloc(size * sizeof(float));

            // Initialize the vector with some values
            for (int i = 0; i < size; i++) {
                vector[i] = (float)i;
            }

            // Measure the start time
            clock_t start = clock();

            // Perform vector scalar addition
            vector_scalar_addition(vector, scalar, size, num_threads[t]);

            // Measure the end time
            clock_t end = clock();

            // Calculate the time taken
            double time_taken = ((double)(end - start)) / CLOCKS_PER_SEC;

            printf("Vector size: %d, Time taken: %f seconds\n", size, time_taken);

            // Clean up
            free(vector);
        }
    }

    return 0;
}

File: assignment2/a.py
import matplotlib.pyplot as plt

# Formatted data for plotting
performance_data = {
    1: [
        (1000, 0.000004),
        (10000, 0.000018),
        (100000, 0.000174),
        (1000000, 0.001751),
        (10000000, 0.017785)
    ],
    2: [
        (1000, 0.000106),
        (10000, 0.000012),
        (100000, 0.000087),
        (1000000, 0.000870),
        (10000000, 0.014493)
    ],
    4: [
        (1000, 0.000152),
        (10000, 0.000007),
        (100000, 0.000047),
        (1000000, 0.008368),
        (10000000, 0.016322)
    ],
    8: [
        (1000, 0.000114),
        (10000, 0.000005),
        (100000, 0.000034),
        (1000000, 0.007984),
        (10000000, 0.049220)
    ]
}

# Extracting the unique vector sizes
vector_sizes = sorted({size for results in performance_data.values() for size, _ in results})

# Plot 1: Performance vs. Vector Size for each thread count
plt.figure(figsize=(10, 6))

for num_threads, results in performance_data.items():
    sizes = [size for size, _ in results]
    times = [time for _, time in results]
    plt.plot(sizes, times, marker='o', label=f'{num_threads} threads')

plt.title('Performance of Vector Scalar Addition with OpenMP')
plt.xlabel('Vector Size')
plt.ylabel('Time Taken (seconds)')
plt.xscale('log')  # Use logarithmic scale for vector size
plt.yscale('log')  # Use logarithmic scale for time taken
plt.legend()
plt.grid(True, ls="--")
plt.savefig('performance_vs_vector_size.png')  # Save the plot as an image
plt.close()

# Plot 2: Thread-Wise Performance for each vector size
plt.figure(figsize=(10, 6))

for size in vector_sizes:
    thread_counts = []
    times = []
    for num_threads, results in performance_data.items():
        for vector_size, time_taken in results:
            if vector_size == size:
                thread_counts.append(num_threads)
                times.append(time_taken)
                break
    plt.plot(thread_counts, times, marker='o', label=f'Vector Size {size}')

plt.title('Thread-Wise Performance of Vector Scalar Addition with OpenMP')
plt.xlabel('Number of Threads')
plt.ylabel('Time Taken (seconds)')
plt.yscale('log')  # Use logarithmic scale for time taken
plt.xticks(sorted(performance_data.keys()))  # Ensure the thread counts are correctly labeled
plt.legend()
plt.grid(True, ls="--")
plt.savefig('thread_wise_performance.png')  # Save the plot as an image
plt.close()
